[fix] Improve csv upload functionality

villebro opened the pull request:
### CATEGORY

Choose one

- [x] Bug Fix
- [x] Enhancement (new features, refinement)
- [x] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Currently all engines except Hive try to create table metadata when uploading CSVs. However, the CSV upload feature throws an exception when uploading using either `replace` or `append` when table metadata is present, despite the upload having taken place. This PR changes the following:
- If table metadata is present and the CSV upload is successful, the old metadata is updated to reflect the new table contents.
- Table metadata is now created for Hive like other engines when uploading CSVs.
- The extension `tsv` added to `ALLOWED_EXTENSIONS` and implemented in `forms.py`, as I feel it is nowadays a commonly accepted extension for tab separated csv files.
- Fixes to i18n CSV messages that were previously hardcoded, i.e. didn't work.
- New integration tests added to test all common cases.
- Refactor: parts of the CSV uploading logic had to be moved from `db_engine_specs` to `database/views` as they would otherwise have caused circular imports.

### TEST PLAN
CI + Tested all possible combinations locally.

### ADDITIONAL INFORMATION
- [x] Has associated issue: closes #8451
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@pengyejun @willbarrett @mistercrunch @john-bodley @betodealmeida

villebro committed:
[fix] csv upload when table metadata present

villebro committed:
Remove table from hive spec

villebro committed:
Move upload before table metadata creation

villebro committed:
Refine upload logic, dd unit tests and fix translations

villebro committed:
Use ALLOWED_EXTENSIONS from config

willbarrett commented on superset/views/database/forms.py:
```diff
@@ -90,7 +90,10 @@ def at_least_one_schema_is_allowed(database):
-        validators=[FileRequired(), FileAllowed(["csv"], _("CSV Files Only!"))],
+        validators=[
+            FileRequired(),
+            FileAllowed(config["ALLOWED_EXTENSIONS"], _("CSV Files Only!")),
```
Now that we're allowing TSV files as well, should this error message be updated?

villebro replied:
Good point; the supported extensions are now included in the error message. I also made the extension checking case-insensitive in `db_engine_specs/base.py` (guessing someone will trip on an upper-case CSV/TSV at some point).

willbarrett commented on superset/views/database/views.py:
```diff
@@ -102,10 +102,10 @@ def form_post(self, form):
-                    database.database_name, schema_name
-                )
+                'Database "%(database_name)s" schema "%(schema_name)s" '
+                "is not allowed for csv uploads. Please contact Superset Admin",
```
I think we're missing a possessive - should this be "Please contact your Superset admin"?

villebro replied:
FIxed that one and found another one at the same time. The other one had exclamation marks which I didn't change now; for uniformity it would be to decide on error grammar/formatting at some point.

villebro committed:
Address review comments

willbarrett commented on superset/views/database/forms.py:
```diff
@@ -92,7 +92,13 @@ def at_least_one_schema_is_allowed(database):
+            FileAllowed(
+                config["ALLOWED_EXTENSIONS"],
+                _(
+                    "Only following file extensions allowed: %(allowed_extensions)s",
```
Grammar nit: perhaps "Only the following file extensions are allowed:"

villebro committed:
Fix error message grammar

villebro commented:
@willbarrett are you ok with the latest changes? @mistercrunch @dpgaspar comments warmly welcomed!

willbarrett commented on superset/db_engine_specs/hive.py:
```diff
@@ -98,7 +98,7 @@ def fetch_data(cls, cursor, limit: int) -> List[Tuple]:
 
     @classmethod
-    def create_table_from_csv(cls, form, table):  # pylint: disable=too-many-locals
+    def create_table_from_csv(cls, form):  # pylint: disable=too-many-locals
```
This change will make the API for the Hive db_engine_spec diverge further from `base`. It would behoove us to keep them in sync.

villebro replied:
Based on my understanding of the `HiveEngineSpec.create_table_from_csv`, this actually aligns the functionality with that of `BaseEngineSpec`, as previously the Hive spec didn't create any metadata for new tables (all the others did). After this change metadata creation takes place outside of the engine spec, and works similarly for all engines, i.e. metadata is created in `database/views.py`, and only uploading of actual data is handled by the engine specs.

willbarrett replied:
Hey Ville - the implementation is fine here, I'm only referring to the removal of the `table` argument from the method signature. Sorry, I should have been more specific.

villebro replied:
@willbarrett I have tried to make sure the API remains compatible, i.e. both `BaseEngineSpec` ja `HiveEngineSpec` now only take `form` as additional parameters (I did forget to add the `None` return type to Hive, though). So there shouldn't be any discrepancy between them. Or did I misunderstand something?

willbarrett commented on superset/views/database/views.py:
```diff
@@ -117,32 +117,58 @@ def form_post(self, form):
-            table.database = form.data.get("con")
-            table.database_id = table.database.id
-            table.database.db_engine_spec.create_table_from_csv(form, table)
+            table_name = form.name.data
```
This is:
1. Becoming a large try/except block
2. Becoming a large method

Can we break up this logic somehow?

villebro replied:
I took a second look at this method, and I agree, it's becoming rather sizeable and difficult to maintain. I did consider refactoring this more, but decided against it, as there was already quite a few big changes happening (I already threw in a few integration tests to feel less guilty 
😆). I don't mind taking another stab at cleaning this up, but would prefer to do it in one or even two separate PRs, as I believe it will require a fairly big effort that falls beyond the scope of this fix. However, if there are strong feelings to the contrary I am happy to take another look at refactoring this more.

willbarrett replied:
Yes, I'm OK deferring the refactor here. Thanks for the response :)

willbarrett reviewed (commented):
A couple more comments on the implementation.

dpgaspar commented on superset/views/database/views.py:
```diff
@@ -117,32 +117,58 @@ def form_post(self, form):
+                    schema=form.schema.data,
+                    database_id=database.id,
+                )
+                .first()
```
We can use `one_or_none` instead

dpgaspar reviewed (approved):
Just a small comment, but it LGTM

villebro committed:
Add return type to hive csv upload and replace first with one_or_none

villebro commented:
Thanks @willbarrett and @dpgaspar for additional comments; I added a return type to the Hive upload method and replaced `first()` with `one_or_none()` in `database/views.py`, I think this should be ok now.
